src/yaml/from_yaml.cr

```diff
@@ -22,39 +22,91 @@ end
+  if YAML::TRUE_VALUES.includes? value
+    true
+  elsif YAML::FALSE_VALUES.includes? value
+    false
+  else
+    raise YAML::ParseException.new("Expected boolean, not '#{value}'", *location)
+  end
 end
 
-{% for type in %w(Int8 Int16 Int32 Int64 UInt8 UInt16 UInt32 UInt64) %}
-  def {{type.id}}.new(pull : YAML::PullParser)
+{% for bits in [8, 16, 32, 64] %}
+  {% for type in %w(Int UInt) %}
+    def {{type.id}}{{bits}}.new(pull : YAML::PullParser)
+      location = pull.location
+      value = pull.read_plain_scalar
+
+      number = {% if type == "UInt" %}
+        value.to_u{{bits}}?(underscore: false, prefix: true)
+      {% else %}
+        value.to_i{{bits}}?(underscore: false, prefix: true)
+      {% end %}
+      return number if number
+
+      raise YAML::ParseException.new("Invalid {{type.id}}{{bits}} number", *location)
+    end
+  {% end %}
+{% end %}
+
+{% for bits in [32, 64] %}
+  def Float{{bits}}.new(pull : YAML::PullParser)
     location = pull.location
-    begin
-      {{type.id}}.new(pull.read_scalar)
-    rescue ex
-      raise YAML::ParseException.new(ex.message.not_nil!, *location)
+    value = pull.read_plain_scalar
+    if float = value.to_f{{bits}}?
+      float
+    elsif YAML::INFINITY_VALUES.includes? value.lchop('+')
+      INFINITY
+    elsif value[0]? == '-' && YAML::INFINITY_VALUES.includes?(value.lchop('-'))
+      -INFINITY
+    elsif YAML::NAN_VALUES.includes? value
+      NAN
+    else
+      raise YAML::ParseException.new("Invalid Float{{bits}} number", *location)
     end
   end
 {% end %}
 
-def String.new(pull : YAML::PullParser)
-  pull.read_scalar
-end
-
-def Float32.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f32
+# TODO: Implement a Time parser that supports all the YAML formats
+def Time.new(pull : YAML::PullParser)
+  location = pull.location
+  begin
+    value = pull.read_plain_scalar
+    Time::Format::ISO_8601_DATE_TIME.parse(value)
+  rescue ex : Time::Format::Error
+    raise YAML::ParseException.new("Could not parse time from '#{value}'", *location)
+  end
 end
 
-def Float64.new(pull : YAML::PullParser)
-  pull.read_scalar.to_f64
+def String.new(pull : YAML::PullParser)
+  location = pull.location
+  return pull.read_scalar unless pull.data.scalar.style == LibYAML::ScalarStyle::PLAIN
+  pull.read_plain_scalar.tap do |value|
+    reserved = YAML::NULL_VALUES.includes?(value) ||
+               YAML::BOOL_VALUES.includes?(value) ||
+               YAML::INFINITY_VALUES.includes?(value.lchop('+')) ||
+               YAML::INFINITY_VALUES.includes?(value.lchop('-')) ||
+               YAML::NAN_VALUES.includes?(value) ||
+               value.to_i64?(underscore: true, prefix: true) ||
+               value.to_f64? ||
+               begin
+                 Time::Format::ISO_8601_DATE_TIME.parse(value)
+               rescue Time::Format::Error
+               end
+
+    raise YAML::ParseException.new("Expected string, not time", *location) if reserved
```

Comment:
IMO `, not time` part of exception message should be either removed or replaced by more accurate statement.

Reply:
@Sija such as?

Reply:
going to do:

```crystal
raise YAML::ParseException.new(%(Expected string, not "#{value}"), *location) if reserved
```